## Part 1 Exercise 4.9

In [2]:
#importing libraries and data

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [46]:
#shortcut for importing data files
path = r'C:\\Users\\anaar\\OneDrive\\Documents\\03-2024-Instacart Basket Analysis'

#importing the pkl df orders_products_combined from 4.8
ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined_4.8.pkl'))

In [5]:
#importing the customer data
customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [7]:
#checking for customer df
customers.shape
customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

## 4. Wrangle the df customers

In [17]:
#renaming columns
customers.rename(columns={'First Name': 'first_name', 'Surnam': 'last_name', 'State': 'state', 'Gender': 'gender', 'Age': 'age'}, inplace=True)

In [19]:
#checking for output
customers.head(10)

user_id first_name last_name  gender           state  age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

I decided to keep all the columns 

## 5. data quality check

In [20]:
customers.describe()

user_id            age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [21]:
#finding missing values
customers.isnull().sum()

user_id             0
first_name      11259
last_name           0
gender              0
state               0
age                 0
date_joined         0
n_dependants        0
fam_status          0
income              0
dtype: int64

In [24]:
#creating a subset for missing first_names
customers_nan = customers[customers['first_name'].isnull() == True]

In [25]:
customers_nan

user_id first_name last_name  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        n_dependants fam_status  income  
53                 2    married   41709  
73                 0     single   82518  
82                 3    married  117099  
99                 1    married  155673  
105                3    married  151819  
...              ...        ...     ...  
206038             3    married   87783  
206044             2    married  108488  
206090             1    married   45275  
206162             0     single   56325  
206171             2    married   59222  

[11259 rows x 10 columns]

These 11259 missing first names represent only 5% of the total data so I decided to keep them as they are.

In [28]:
#checking for duplicates
customers_dups = customers[customers.duplicated()]

In [30]:
#checking for output
customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, n_dependants, fam_status, income]
Index: []

There are no duplicates in the customers df

In [31]:
#Checking the data type of each column in the df_ords dataframe
print(customers.dtypes)

user_id          int64
first_name      object
last_name       object
gender          object
state           object
age              int64
date_joined     object
n_dependants     int64
fam_status      object
income           int64
dtype: object


In [32]:
customers.describe()

user_id            age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [ ]:
## Mixed data type
- checking at the number columns for the min and max to make sense. 
- For the user_id it makes sense that it start with 1 and finished the 206209 the total number of rows in this df
- The age also makes sense, starts at 18, and the max is 81
- n_dependants also makes sense, it might be that instacart fixed the max number of dependants to 3
- income column has the same number for mean and 50% quantile which makes me think this column is also correct 

I think it only makes sense to change the data type of date_joined from object to datetime

In [36]:
#changing the data type of date_joined
customers['date_joined'] = pd.to_datetime(customers['date_joined'], format='MM-DD-YYYY')

In [37]:
#checking for output
print(customers.dtypes)

user_id                  int64
first_name              object
last_name               object
gender                  object
state                   object
age                      int64
date_joined     datetime64[ns]
n_dependants             int64
fam_status              object
income                   int64
dtype: object


In [39]:
#checking the final head
customers.head(10)

user_id first_name last_name  gender           state  age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48  2017-01-01   
1    33890   Patricia      Hart  Female      New Mexico   36  2017-01-01   
2    65803    Kenneth    Farley    Male           Idaho   35  2017-01-01   
3   125935   Michelle     Hicks  Female            Iowa   40  2017-01-01   
4   130797        Ann   Gilmore  Female        Maryland   26  2017-01-01   
5   133128    Cynthia     Noble  Female        Kentucky   43  2017-01-01   
6   152052      Chris    Walton    Male         Montana   20  2017-01-01   
7   168851     Joseph   Hickman    Male  South Carolina   30  2017-01-01   
8    69965     Jeremy      Vang    Male           Texas   47  2017-01-01   
9    82820      Shawn     Chung    Male        Virginia   26  2017-01-01   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374  
5             2    married   49643  
6             0     single   61746  
7             0     single   63712  
8             1    married  162432  
9             2    married   32072

## 6. Combining the dataframes

In [47]:
#looking a comun key
ords_prods_combined.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   

   ... merge_indicator     busiest_day    busiest_days  \
0  ...            both  Regularly busy  Regularly busy   
1  ...            both  Regularly busy    Slowest busy   
2  ...            both  Regularly busy    Slowest busy   
3  ...            both      Least busy    Slowest busy   
4  ...            both      Least busy    Slowest busy   
5  ...            both  Regularly busy  Regularly busy   
6  ...            both  Regularly busy    Busiest days   
7  ...            both  Regularly busy    Busiest days   
8  ...            both  Regularly busy    Busiest days   
9  ...            both      Least busy    Slowest busy   

                            busiest_period_of_day max_order  loyalty_flag  \
0  Average orders in hours: 7,8,17,18,19,20,21,22        10  New customer   
1  Average orders in hours: 7,8,17,18,19,20,21,22        10  New customer   
2    Most orders in hours: 9,10,11,12,13,14,15,16        10  New customer   
3  Average orders in hours: 7,8,17,18,19,20,21,22        10  New customer   
4    Most orders in hours: 9,10,11,12,13,14,15,16        10  New customer   
5  Average orders in hours: 7,8,17,18,19,20,21,22        10  New customer   
6    Most orders in hours: 9,10,11,12,13,14,15,16        10  New customer   
7    Most orders in hours: 9,10,11,12,13,14,15,16        10  New customer   
8    Most orders in hours: 9,10,11,12,13,14,15,16        10  New customer   
9  Average orders in hours: 7,8,17,18,19,20,21,22        10  New customer   

  avg_spend spender_flag  order_frequency         frequency_flag  
0  6.367797  Low spender             20.5  Non-frequent customer  
1  6.367797  Low spender             20.5  Non-frequent customer  
2  6.367797  Low spender             20.5  Non-frequent customer  
3  6.367797  Low spender             20.5  Non-frequent customer  
4  6.367797  Low spender             20.5  Non-frequent customer  
5  6.367797  Low spender             20.5  Non-frequent customer  
6  6.367797  Low spender             20.5  Non-frequent customer  
7  6.367797  Low spender             20.5  Non-frequent customer  
8  6.367797  Low spender             20.5  Non-frequent customer  
9  6.367797  Low spender             20.5  Non-frequent customer  

[10 rows x 24 columns]

In [48]:
ords_prods_combined.shape

(32404859, 24)

In [42]:
customers.shape

(206209, 10)

In [50]:
#checking for data type in ords_prods_combined
print(ords_prods_combined.dtypes)

order_id                     int64
user_id                      int64
order_number                 int64
orders_day_of_week           int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
_merge                    category
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
merge_indicator           category
busiest_day                 object
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
avg_spend                  float64
spender_flag                object
order_frequency            float64
frequency_flag              object
dtype: object


In [54]:
#new dataframe using the user_id as a key
df_combined = pd.merge(ords_prods_combined, customers, on='user_id', indicator='merge_indicator_4.9')

In [55]:
df_combined.shape

(32404859, 34)

In [56]:
df_combined.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN         196                  1          0   both   
1                    15.0         196                  1          1   both   
2                    21.0         196                  1          1   both   
3                    29.0         196                  1          1   both   
4                    28.0         196                  1          1   both   
5                    19.0         196                  1          1   both   
6                    20.0         196                  1          1   both   
7                    14.0         196                  2          1   both   
8                     0.0         196                  4          1   both   
9                    30.0         196                  1          1   both   

   ... first_name  last_name  gender    state age date_joined n_dependants  \
0  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
1  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
2  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
3  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
4  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
5  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
6  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
7  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
8  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   
9  ...      Linda     Nguyen  Female  Alabama  31  2019-02-17            3   

  fam_status  income merge_indicator_4.9  
0    married   40423                both  
1    married   40423                both  
2    married   40423                both  
3    married   40423                both  
4    married   40423                both  
5    married   40423                both  
6    married   40423                both  
7    married   40423                both  
8    married   40423                both  
9    married   40423                both  

[10 rows x 34 columns]

## Exporting the combined df

In [58]:
df_combined.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'df_combined_4.9_part1.pkl'))